lib/src/socket_connector.dart

```diff
@@ -136,31 +140,34 @@ class SocketConnector {
           'Added connection. There are now ${connections.length} connections.'));
 
       for (final side in [thisSide, thisSide.farSide!]) {
-        unawaited(side.socket.done
-            .then((v) => _destroySide(side))
-            .catchError((err) => _destroySide(side)));
         if (side.transformer != null) {
           // transformer is there to transform data originating FROM its side
+          // transformer's output will write to the SOCKET on the far side
           StreamController<Uint8List> sc = StreamController<Uint8List>();
           side.farSide!.sink = sc;
           Stream<List<int>> transformed = side.transformer!(sc.stream);
-          transformed.listen((data) {
-            try {
-              if (side.farSide!.state == SideState.open) {
+          transformed.listen(
+            (data) {
+              try {
                 side.farSide!.socket.add(data);
-              } else {
-                throw StateError(
-                    'Will not write to side ${side.farSide!.name} as its state is ${side.farSide!.state}');
+                side.farSide!.sent += data.length;
+                if (side.state == SideState.closed &&
+                    side.rcvd == side.farSide!.sent) {
+                  _closeSide(side.farSide!);
+                }
```

Comment:
Re the above code: Since the transformer stream may not have written everything to the other side's socket before this side's socket closes, we keep track of number of bytes sent, and, if this side has been closed, will call _closeSide on the far side which will result in both side's sockets being cleaned up correctly